src/application/proxy.ts

```diff
@@ -34,6 +53,39 @@ export default class WindowProxy {
       window.location.origin
     );
   }
+
+  on(type: MarinaEventType, callback: (payload: any) => void): EventListenerID {
+    const id = makeid(8);
+    this.addEventListener(type, { id, listener: callback });
+    return id;
+  }
+
+  off(id: EventListenerID) {
+    this.removeEventListener(id);
+  }
+
+  // start the window listner for a given marina event type
+  private startWindowListener(type: MarinaEventType) {
+    window.addEventListener(`marina_event_${type.toLowerCase()}`, (event: Event) => {
+      const payload = (event as CustomEvent).detail;
+      for (const eventListener of this.eventListeners[type]) {
+        eventListener.listener(payload);
+      }
+    });
+  }
+
+  private addEventListener(type: MarinaEventType, listener: MarinaEventListener) {
+    this.eventListeners[type].push(listener);
+    if (this.eventListeners[type].length === 1) this.startWindowListener(type);
```

Comment:
This way we can add a single listener per event type. A developer may want to register many different callbacks for an event

Reply:
no, this line only starts the window's listener the first time we add a callback. but several callbacks are allowed for a same event type here :thinking: